src/clean.c

```diff
@@ -189,13 +189,51 @@ static bool is_all_digits(const char *s)
 	return true;
 }
 
+static bool is_all_xdigits(const char *s)
+{
+	for (; *s; s++) {
+		if (!isxdigit(*s)) {
+			return false;
+		}
+	}
+	return true;
+}
+
 static bool is_manifest(const char UNUSED_PARAM *dir, const struct dirent *entry)
 {
 	static const char prefix[] = "Manifest.";
 	static const size_t prefix_len = sizeof(prefix) - 1;
 	return !strncmp(entry->d_name, prefix, prefix_len);
 }
 
+static bool is_hashed_manifest(const char UNUSED_PARAM *dir, const struct dirent *entry)
+{
+	/* check that this has the manifest prefix */
+	if (!is_manifest(dir, entry)) {
+		return false;
+	}
+
+	int i;
+	char *ename, *tmp = NULL;
+	tmp = ename = strdup_or_die(entry->d_name);
+
+	/* check for the correct number of '.' characters
+	 * (i.e. Manifest.bundlename.hashvalue). Note, this will not clean up
+	 * iterative manifests (Manifest.bundlename.I) because I is not a valid
+	 * xdigit.
+	 */
+	for (i = 0; ename[i]; ename[i] == '.' ? i++ : *ename++) {
```

Comment:
you could use entry->d_name here so you won't need to duplicate memory. *ename++ looks wrong too.
I would use strchr for that and then you can get the pointer to inform is_all_xdigits so you won't need to look for '.' again in line 234

Reply:
why does *ename++ look wrong to you? I am marching through the string by updating the pointer to it. I copy the memory because I can't (and wouldn't want to) modify the `const char *entry->d_name`

Reply:
It is wrong because you are not moving the pointer. *ename++ is changing the value of the memory pointed by ename, so you are updating the string and not moving the pointer forward. that's why you need to move i forward too. If you use a string "012345" to this loop the string will be "612345" after that.

If you are doing a read only operation you never need to duplicate the string. Just move the pointer: 

    char *tmp;
    for (tmp = entry->d_name; tmp; tmp++)

But you don't need to do that. As you are already looking for the first '.' in the string you just need to look for the second one. You can do this:

    #define MANIFEST_PREFIX "Manifest."
    
    suffix = strchr(entry->d_name + sizeof(MANIFEST_PREFIX), '.'); // Look for next '.' after Manifest.
    return suffix && is_all_xdigits(suffix + 1); // if '.' was found check if all digits after '.' are xdigits

Reply:
ops, my bad. the comment about  *ename++ is incorrect. It's not updating the content. but anyway, I don't get why moving i and ename foward.

Reply:
Ok, now I got that. It's to count the number of occurrences at the same time you're moving forward. haha. If you want to keep this you can, but add some comment about that.